### DISCUSSION TITLE
Setting up Rust LSP

### DISCUSSION THREAD

**Discussion by NeotenicPrimate** [helix-editor/helix#4737] [Q&A] [answered]
I have followed the instructions layed out here [Languages](https://docs.helix-editor.com/languages.html), here [rust-analyzer Language Server Binary](https://rust-analyzer.github.io/manual.html#rust-analyzer-language-server-binary) and here [How to install the default language servers](https://github.com/helix-editor/helix/wiki/How-to-install-the-default-language-servers#rust_analyzer).

I have placed the binary for the language server in `$PATH` as indicated by:

```
hx --health rust

Configured language server: rust-analyzer
Binary for language server: /Users/MyName/.cargo/bin/rust-analyzer
Configured debug adapter: lldb-vscode
Binary for debug adapter: Not found in $PATH
Highlight queries: ✓
Textobject queries: ✓
Indent queries: ✓
```

My `languages.toml` looks like this:

```
[[language]]
name = "rust"

[language.config]
checkOnSave = { command = "clippy" }

[[grammar]]
name = "rust"
source = { git = "https://github.com/tree-sitter/tree-sitter-rust", rev = "0431a2c60828731f27491ee9fdefe25e250ce9c9" }
```

Currently when I `Space-k` Helix returns `Language server not active for current buff`.

My goal is to get the following to work
- Inline compile time errors
- Type hints
- Auto-complete suggestions

**Comment by CptPotato** [comment]
How did you create your rust project? Did you use `cargo new ...`?
You need a project root (like `Cargo.toml`) for the LSP to work.

If that's the case and it still doesn't work you can check/provide the log file for more information (you can use `hx --help` to see where the log is located).

**Reply by NeotenicPrimate** [reply]
Yes I tried both opening an existing project as well as creating a new one (`cargo new ...`).
Upon inspecting the logs, as you suggested, it seems that the binary is not reachable.

`11 2022-11-07T20:14:19.929 helix_view::editor [ERROR] Failed to initialize the LSP for 'source.rust' { cannot find binary path }`

(There are many other errors, but I am guessing they all stem from the same problem)

I put the `exec` binary in `$PATH`, specifically `/Users/MyName/.cargo/bin`.

<img width="248" alt="Screen Shot 2022-11-14 at 8 50 55 PM" src="https://user-images.githubusercontent.com/67292972/201829302-6dc866ce-65e0-4590-adbb-70143f8569bb.png">

**Comment by NeotenicPrimate** [comment, marked as answer]
Rather than trying to do it manually, all you need to do is run:

`sudo ln -s $(rustup which rust-analyzer ) /usr/local/bin/rust-analyzer`

As indicated [here](https://github.com/helix-editor/helix/wiki/How-to-install-the-default-language-servers#rust_analyzer).

**Reply by nuke-web3** [reply]
Strangely for me: I got health of:
```
rust                   ✓ rust-analyzer        ✓ lldb-vscode          ✓                      ✓                      ✓
```
but also the same `Space-k` Helix returns `Language server not active for current buff.` error, my resolution was a dumb one, I didn't have RA actually installed for _my toolchain_, just needed:
```sh
rustup component add rust-analyzer
```

**Comment by CGMossa** [comment]
I still don't know how to get the two things in the path on Windows.

**Reply by kirawi** [reply]
https://stackoverflow.com/questions/1618280/where-can-i-set-path-to-make-exe-on-windows

**Comment by ghost** [comment]
My C and python can use lsp greatly, but rust lsp is broken.
I set languages.toml the same as NeotenicPrimate's,and have Binary for language server: /usr/local/bin/rust-analyzer, but still can't work with lsp,this is the error:
```
helix_lsp::transport [ERROR] err <- "[ERROR rust_analyzer::lsp_utils] rust-analyzer failed to load workspace: Failed to find sysroot for Cargo.toml file /home/username/tools/helix/Cargo.toml. Is rust-src installed?: can't load standard library from sysroot\n"
helix_lsp::transport [ERROR] err <- "/usr\n"
helix_lsp::transport [ERROR] err <- "(discovered via `rustc --print sysroot`)\n"
helix_lsp::transport [ERROR] err <- "try installing the Rust source the same way you installed rustc\n"
```
`rustc --print sysroot` shows: /usr/
why rust-analyzer failed to load workspace?
I am newbie to rust, is Cargo.toml matter with this problem.

**Reply by ghost** [reply]
I saw the same issue in rust-analyzer, I tried rm rust by `pacman -R rust`, and it work well to me.  it should be caused by the packagemanager.

**Comment by dmgolembiowski** [comment]
For anyone encountering this problem in 2025, this fixed my lsp integration:

```sh
rustup self update
rustup self upgrade-data
rustup component rust-analyzer
```
